### PR TITLE
Bios date location changed in W10

### PR DESCRIPTION
**Pull request by fschenckel**
Bios date location in registry seems to have changed at least since W10

### CODE REVIEW

**Comment by ronaldhoek** [comment]
Sorry, you'll need to add some backward compatibility: detect if OS is Win10 (or higher?) and then use the corresponding registry keys based on the detection (old vs new)

**Comment by fschenckel** [comment]
Hi,

The problem is I couldn't get any information regarding both keys... The only thing I'm sure is that the new key I propose to use is already existing in Win7.
The official version is failing on W10 and higher.

I could eventually check if OS > Win7 ?

Regards

Frédéric SCHENCKEL

* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
cab technologies
za Nord du val de Moder
67350 Niedermodern
Tél: + 33 3 88 72 25 01 / Fax: +33 3 88 72 25 02
Notre site en Français: www.cab-technologies.fr<http://www.cab-technologies.fr/>
Adresse support: ***@***.******@***.***>
* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

De : Ronald Hoek ***@***.***>
Envoyé : mercredi 30 novembre 2022 15:29
À : project-jedi/jcl ***@***.***>
Cc : Frédéric SCHENCKEL ***@***.***>; Author ***@***.***>
Objet : Re: [project-jedi/jcl] Bios date location changed in W10 (PR #122)


Sorry, you'll need to add some backward compatibility: detect if OS is Win10 (or higher?) and then use the corresponding registry keys based on the detection (old vs new)

-

**Comment by ronaldhoek** [comment]
Ok, don't have any knowledge about the keys either.

Another idea would be to first use the new key and when no value is found, use the old key?

**Comment by fschenckel** [comment]
OK, yes I can do it this way:
1) Check for the 'new' key (W10 and newer => Most probability today)
2) if not existing use the key as it is done today, so we will have the same behavior regarding exception and so on...

**Comment by ronaldhoek** [comment]
@obones can you merge this - I can see the date a the new location (in my Win10 setup) and no longer at the old location - so regarding backward compatibility it looks fine.

File: jcl/source/common/JclSysInfo.pas [hunk 1]
```diff
@@ -2606,8 +2606,10 @@ function GetBIOSExtendedInfo: string;
 { TODO : the date string can be e.g. 00/00/00 }
 function GetBIOSDate: TDateTime;
 const
-  WinNT_REG_PATH = 'HARDWARE\DESCRIPTION\System';
-  WinNT_REG_KEY  = 'SystemBiosDate';
+  WinNT_REG_PATH_NEW = 'HARDWARE\DESCRIPTION\System\BIOS';
+  WinNT_REG_PATH_OLD = 'HARDWARE\DESCRIPTION\System';
+  WinNT_REG_KEY_NEW  = 'BIOSReleaseDate';
+  WinNT_REG_KEY_OLD  = 'SystemBiosDate';
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2612]: I dislike the "new"/"old" naming as it doesn't convey any meaning as to when the change happened. And what if it changes again? Will it then be "NEW_NEW" or "NEW_2"?
I'd rather keep `WinNT_REG_PATH` and `WinNT_REG_KEY` introduce `WIN10_REG_PATH` and `WIN10_REG_KEY` both declared together right after the `const` line

File: jcl/source/common/JclSysInfo.pas [hunk 2]
```diff
@@ -2619,8 +2621,16 @@ function GetBIOSDate: TDateTime;
   RegSeparator: Char;
   {$ENDIF ~RTL150_UP}
 begin
-  if IsWinNT then
-    RegStr := RegReadString(HKEY_LOCAL_MACHINE, WinNT_REG_PATH, WinNT_REG_KEY)
+  if IsWinNT then begin
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2624]: The coding style guide mandates that `begin` be placed on its own line.
Also, if a branch of the `if`/`else` statement requires a `begin`/`end` block, then the other part must have it also.

File: jcl/source/common/JclSysInfo.pas [hunk 3]
```diff
@@ -2619,8 +2621,16 @@ function GetBIOSDate: TDateTime;
...
+    {fs 14/12/2022: location of the Bios date seems to have changed on newer systems (From windows 10 ?)
+    The new location seems to exist since a while, but older location disappeared on newer OS}
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2626]: No need for the initials and the date. Also, two lines comments like this one should not use the curly braces style but rather the two slashes style, repeated on both lines

File: jcl/source/common/JclSysInfo.pas [hunk 4]
```diff
@@ -2619,8 +2621,16 @@ function GetBIOSDate: TDateTime;
...
+    if RegValueExists(HKEY_LOCAL_MACHINE, WinNT_REG_PATH_NEW, WinNT_REG_KEY_NEW) then
+      {New location}
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2628]: Please remove this line

File: jcl/source/common/JclSysInfo.pas [hunk 5]
```diff
@@ -2619,8 +2621,16 @@ function GetBIOSDate: TDateTime;
...
+      RegStr := RegReadString(HKEY_LOCAL_MACHINE, WinNT_REG_PATH_NEW, WinNT_REG_KEY_NEW)
+    else
+      {Old location}
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2631]: Please remove this line

File: jcl/source/common/JclSysInfo.pas [hunk 6]
```diff
@@ -2620,7 +2622,14 @@ function GetBIOSDate: TDateTime;
...
+  begin
+    // location of the Bios date seems to have changed on newer systems (From windows 10 ?)
+    // The new location seems to exist since a while, but older location disappeared on newer OS
+    if RegValueExists(HKEY_LOCAL_MACHINE, WIN10_REG_PATH, WIN10_REG_KEY) then
+      RegStr := RegReadString(HKEY_LOCAL_MACHINE, WIN10_REG_PATH, WIN10_REG_KEY)
+    else
+      RegStr := RegReadString(HKEY_LOCAL_MACHINE, WinNT_REG_PATH, WinNT_REG_KEY);
+  end
   else
     RegStr := RegReadString(HKEY_LOCAL_MACHINE, Win9x_REG_PATH, Win9x_REG_KEY);
```
Comment (obones) [inline jcl/source/common/JclSysInfo.pas:2634]: The coding style mandates that if a branch of the if/else statement requires a begin/end block, then the other part must have it also.
So please add a begin/end statement around this one

**Review by obones — CHANGES_REQUESTED** [review]

**Pull request merged**